Enhance infrastructure docs, add perf testing results (#669)
vegaprotocol/documentation: docs/node-operators/requirements/data-node-infrastructure.md

---
sidebar_position: 2
title: Data node infrastructure
sidebar_label: Data node infrastructure
vega_network: TESTNET
hide_title: false
---
import NetworkParameter from '@site/src/components/NetworkParameter';
import Topic from '/docs/topics/_topic-data-nodes.mdx';

<Topic />

This page covers the minimum and recommended hardware requirements, and the software requirements for running a data node connecting to the Vega network.

The best hardware may vary depending on how you set up your node and database service, and how you intend to use the data node.

Data node operators who are running a data node to support their own needs can run a leaner setup depending on their own requirements.

Those data node operators running archive nodes, which retain all chain information, will need a more robust setup.

## Hardware recommendations

### System requirements
| Resource    | Minimum     | Recommended |
| ----------- | ----------- | ----------- |

> | CPU | 6 cores at >= 3Ghz | 12 cores at >= 4Ghz|

Reviewer: **question (non-blocking)**:  Do we need to mention architecture here?

| RAM   | 16GB            | 32GB        |
| Storage   | 1TB NVMe SSD| 2TB NVMe SSD |

The impact of using fewer cores than recommended is that the critical parts (data node and non-validators) will be starved of CPU resources, which will reduce the maximum throughput of the network.

> The impact of having slower cores than recommended (or older cores that have a reduced IPC rate compared to modern Zen3 cores) is that the maximum throughput will be reduced due to the bottlenecks being single threaded.

Reviewer: **question (non-blocking)**: And what's the consequence of this?

```suggestion
The impact of having slower cores than recommended (or older cores that have a reduced IPC rate compared to modern Zen3 cores) is that the maximum throughput will be reduced due to the bottlenecks being single threaded. This could cause your validator to fall behind the others, which could lead to your removal from the validator set.
```

Author: This is the data node page, so this isn't a factor here.

### File system performance testing

> In benchmark testing using the default Ubuntu file system of EXT4 against ZFS with zstd compression turned on, it was determined that ZFS has an impact on performance when using NVMe drives, while the opposite is seen in SATA drives.

Reviewer: **thought (non-blocking)**: I'm not all that familiar with zstd so my initial thought was that linking off to somewhere (e.g. a github page or some zfs docs for enabling this) might be good. But I think that's a step too far in to hand holding.

Reviewer 2: There will be people out there who do not know what EXT4 is as well, but I think if they are considering changing their filesystem from the default they need to go and and do some reading by themselves as they will need to understand how to create new partitions and format them correctly. This would be too complicated for this doc.

|     | NVMe SSD (bpm)  | Service Time | SATA SSD (bpm) | Service Time |
| ----------- | ----------- | ----------- | ----------- |
| Core EXT4 | 1415 | 15% | 851 | 60% |  
| Core ZFS | 1210  | 15% | 886 | 55% |
| Core + Datanode EXT4 | 1216 | 35% | 607 | 90% |
| Core + Datanode ZFS | 993 | 25% | 729 | 65% |

### Cores performance testing 
Benchmark testing shows that ZFS with zstd helps reduce the amount of data stored on the drives, but has a negative impact on the amount of CPU resources required. This is due to the compression/decompression of the data as it is written and read from disk. 

Below are results of replay tests showing how different amounts of available CPU cores change the average block processing rate.

> | Cores | EXT4 (bpm) | ZFS + zstd (bpm) |

Reviewer: **question (non-blocking)**: Beats Per Minute? So music playback is slower with zstd?

| ----------- | ----------- | ----------- |
| 1 | 496 | 326 |
| 2   | 853 | 604 |
| 3   | 1003 | 751 |
| 4   | 1439 | 856 |
| 5   | 1471 | 1014 |
| 6   | 1466 | 1164 |
| 7   |  | 1172 |

## Software requirements

### Operating system 
For production use, we recommend using the Linux binary on Ubuntu as this is the platform used by nodes on Fairground, the Vega testnet, and is the most widely tested so far. 

### Golang

> You'll need 'go' version 1.19.1 or newer.

Reviewer 3: for info this is valid for vega 0.71.5 but not for future release.

Author: So 1.20 will be required, is that on the next minor release?

### Vega Wallet

> You'll need a version of the Vega Wallet that matches with the version of software you're using to install your node. You can get this when you install Vega. 

Reviewer: **nitpick (non-blocking)**: This sentence feels a little bit muddled. But I can't come up with a better suggestion

Author: Yep, it can go and it'll be gone in the next commit.

### Data node PostgreSQL

> After building, to run your data node, you will have to run a PostgreSQL server.

Reviewer: ```suggestion
If you plan to run a data node, you will also need to run a PostgreSQL server.
```
